File: group-anagram/step-1.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        ans = {}
        
        for s in strs:
            arr = [0] * 26
            for l in s:
                arr[ord(l) - 97] += 1
            ans.get(tuple(arr), []).append(s)
        return ans.values()

File: group-anagram-2/step1.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
      unordered_map<string, vector<string>> mp;

      for (const string& s : strs) {
        array<int, 26> count{};
        count.fill(0);

        for (char c : s) {
          count[c - 'a']++;
        }

        string key;
        
        for (int num : count) {
          key += to_string(num);
          key += '#';
        }

        mp[key].push_back(s);
      }

      vector<vector<string>> result;
      result.reserve(mp.size());

      for (auto& [key, group] : mp) {
        result.push_back(std::move(group));
      }
      
      return result;
    }
};
